feat: implement aws-eks-addon datasource

jzhn opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

Add a new datasource, `aws-eks-addon`. It uses similar design as existing `aws-machine-image` and `aws-rds` to use AWS JavaScript SDK to call APIs to fetch AWS EKS addon versions. 

## Context

<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number. -->
<!-- If you're referencing an issue with this pull request, put it in a Markdown list like this: - #issue_number. -->

AWS EKS Addons are Kubernetes data-plane components under EKS's management, like [kube-proxy](https://docs.aws.amazon.com/eks/latest/userguide/managing-kube-proxy.html) and [coredns](https://docs.aws.amazon.com/eks/latest/userguide/managing-coredns.html). The addons are usually semver versioned, and released regularly for new features and security patches. AWS EKS provides API [`DescribeAddonVersions`](https://docs.aws.amazon.com/eks/latest/APIReference/API_DescribeAddonVersions.html) to allow users to query available addon versions per target Kubernetes version.

As platform engineer, I often need to ensure my EKS clusters are running with up-to-date addons. Before this PR, renovate is unable to fetch list of available EKS addon versions that's specific to my cluster version and region. This is a [popular ask](https://github.com/renovatebot/renovate/issues/23410#issuecomment-1877355858) of renovate users. 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository
  - 100% unit test coverage
  - tested end-to-end in my self-hosted setup with a terraform project

<img width="856" alt="image" src="https://github.com/renovatebot/renovate/assets/4455171/aa9fa341-e206-403c-a73c-ee22f75d35c4">

<img width="910" alt="image" src="https://github.com/renovatebot/renovate/assets/4455171/f341375d-82e6-4701-a4ed-8c0f632fc077">


<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

viceice commented on lib/modules/datasource/aws-eks-addon/index.spec.ts:
```diff
@@ -0,0 +1,125 @@
+      expect(res).toBeNull();
+      expect(eksMock.calls()).toHaveLength(1);
+      expect(eksMock.call(0).args).toMatchSnapshot();
+      expect(await eksMock.call(0).returnValue).toMatchSnapshot();
```
please don't use snapshots 

https://github.com/renovatebot/renovate/blob/main/docs/development/best-practices.md#unit-testing

jzhn replied:
Thanks for the review. Replaced the toMatchSnapshot with explicit response assertions.

jzhn commented:
@viceice : can you please take another look? I've addressed your comment and all checks are green. thank you :)

viceice commented on lib/modules/datasource/aws-eks-addon/index.ts:
```diff
@@ -0,0 +1,64 @@
+    packageName: serializedFilter,
+  }: GetReleasesConfig): Promise<ReleaseResult | null> {
+    const filter: EKSAddonsFilter = JSON.parse(serializedFilter);
+    const eksClient = this.getEKSClient(filter);
```
why do we need a new client on each request? this can cause a massive slowdown

jzhn replied:
AWS EKS is a regional service. In practice, we can expect that EKS should make Addon versions consistent across all commercial regions, but in theory they could choose to release a version in selected regions only. So I decided to allow users to optionally specify AWS `region` per EKS addon "package", which would require rebuilding the EKS client per request.

The code is similar to AWS AMI datasource: https://github.com/jzhn/renovate/blob/feat/23410_aws-eks-addon-support/lib/modules/datasource/aws-machine-image/index.ts#L93. 

If we build with the assumption that EKS addon versions are consistent, we could have a singleton EKSClient. but I feel this might cause confusion for users.

viceice replied:
then cache the client per region, so it's reusable.

jzhn replied:
Updated to cache clieint per region + profile.

viceice commented on lib/modules/datasource/aws-eks-addon/index.ts:
```diff
@@ -0,0 +1,64 @@
+        .flatMap((addon) => addon.addonVersions)
+        .map((versionInfo) => ({
+          version: versionInfo!.addonVersion!,
+          isDeprecated: false,
```
remove, default is false

viceice commented on lib/modules/datasource/aws-eks-addon/index.ts:
```diff
@@ -0,0 +1,64 @@
+      releases: addons
+        .flatMap((addon) => addon.addonVersions)
+        .map((versionInfo) => ({
+          version: versionInfo!.addonVersion!,
```
why do we need all those `!`?

jzhn replied:
AWS javascript SDK makes everything nullable even for basic fields that should always exist. I added an extra `filter` to safeguard this (and updated unit tests). Hope it looks better to you :)

viceice replied:
do an additional map and filter after, so no need for any `!`

jzhn replied:
I refactored it a bit, hope it looks better to you.

jzhn commented on lib/modules/datasource/aws-eks-addon/index.spec.ts:
```diff
@@ -0,0 +1,140 @@
+  EKSClient,
+} from '@aws-sdk/client-eks';
+import { mockClient } from 'aws-sdk-client-mock';
+import 'aws-sdk-client-mock-jest';
```
It's to have some additional Jest matchers for assert on AWS api calls. https://www.npmjs.com/package/aws-sdk-client-mock#user-content-jest-matchers

I can replace it with basic jest apis.

viceice replied:
yes, please use basic jest API.

jzhn commented on package.json:
```diff
@@ -310,6 +311,7 @@
     "@typescript-eslint/eslint-plugin": "7.13.0",
     "@typescript-eslint/parser": "7.13.0",
     "aws-sdk-client-mock": "4.0.1",
+    "aws-sdk-client-mock-jest": "4.0.1",
```
replied in https://github.com/renovatebot/renovate/pull/29613#discussion_r1641583735, will be replaced with basic jest apis.

viceice commented on lib/modules/datasource/aws-eks-addon/index.spec.ts:
```diff
@@ -0,0 +1,153 @@
+
+describe('modules/datasource/aws-eks-addon/index', () => {
+  describe('getPkgReleases()', () => {
+    it.each([
```
please use string template syntax

jzhn replied:
Updated, I hope I'm doing what you have in mind, please take a look

viceice commented on lib/modules/datasource/aws-eks-addon/index.spec.ts:
```diff
@@ -0,0 +1,153 @@
+const datasource = AwsEKSAddonDataSource.id;
+const eksMock = mockClient(EKSClient);
+
+const addonVersion1: AddonVersionInfo = {
```
please inline those variables to appropriate scope, that are all only used once. (perf)

viceice commented on lib/modules/datasource/aws-eks-addon/index.ts:
```diff
@@ -0,0 +1,64 @@
+  async getReleases({
+    packageName: serializedFilter,
+  }: GetReleasesConfig): Promise<ReleaseResult | null> {
+    const filter: EKSAddonsFilter = JSON.parse(serializedFilter);
```
use zod Schema to validate filter values

jzhn commented:
@viceice : apologies for the code smells and anti patterns as typescript isn't my daily language. I did try my best to understand and address your comments, hopefully the latest revision looks better to you. cheers.

jzhn commented:
Hi @viceice , it's been a week since I addressed your feedbacks. Can you please take another look or let me know what's preventing another review? Thanks.

cc @rarkins

viceice commented on lib/modules/datasource/aws-eks-addon/index.ts:
```diff
@@ -0,0 +1,72 @@
+
+  @cache({
+    namespace: `datasource-${AwsEKSAddonDataSource.id}`,
+    key: ({ packageName }: GetReleasesConfig) => `getReleases:${packageName}`,
```
🤔 Shouldn't we take `eksAddonsFilter` into account here? Otherwise we can get cache clash when using different filters for same package name.

rarkins commented:
@jzhn please see the code review comment above when you get the time

ahoehma commented:
Any chance to get this soon? 👍

viceice commented on lib/modules/datasource/aws-eks-addon/schema.ts:
```diff
@@ -0,0 +1,10 @@
+  profile: z.string().optional(),
+});
+
+export type EKSAddonsFilter = z.infer<typeof EKSAddonsFilterSchema>;
```
```suggestion
export type EksAddonsFilter = z.infer<typeof EKSAddonsFilterSchema>;
export EksAddonsFilterJson = Json.pipe(EksAddonsFilterSchema);
```
from schema utils

viceice commented on lib/modules/datasource/aws-eks-addon/schema.ts:
```diff
@@ -0,0 +1,10 @@
+import { z } from 'zod';
+
+export const EKSAddonsFilterSchema = z.object({
```
```suggestion
export const EksAddonsFilterSchema = z.object({
```

viceice commented on lib/modules/datasource/aws-eks-addon/index.ts:
```diff
@@ -0,0 +1,72 @@
+  async getReleases({
+    packageName: serializedFilter,
+  }: GetReleasesConfig): Promise<ReleaseResult | null> {
+    const filter: EKSAddonsFilter = EKSAddonsFilterSchema.parse(
+      JSON.parse(serializedFilter),
+    );
```
```suggestion
    const filter = EksAddonsFilterJson.safeParse(serializedFilter);
```
Simplify and validate result

viceice commented on lib/modules/datasource/aws-eks-addon/index.ts:
```diff
@@ -0,0 +1,72 @@
+import { cache } from '../../../util/cache/package/decorator';
+import { Datasource } from '../datasource';
+import type { GetReleasesConfig, ReleaseResult } from '../types';
+import { EKSAddonsFilter, EKSAddonsFilterSchema } from './schema';
```
```suggestion
import { EksAddonsFilterJson } from './schema';
```

viceice reviewed (commented):
use the re-request review button when done

lmayorga1980 commented:
Is the `EksAddonsFilterSchema` should be used as the input for `getEKSClient`? Comparing to `amazon-machine-image` perhaps those should be 2 different schemas

jvdadda commented:
Nice work @jzhn , do you think it is possible to update the PRs with last main ? And re-request a review ? Thanks a lot

viceice commented on lib/modules/datasource/aws-eks-addon/schema.ts:
```diff
@@ -0,0 +1,11 @@
+});
+
+export type EksAddonsFilter = z.infer<typeof EKSAddonsFilterSchema>;
+export EksAddonsFilterJson = Json.pipe(EksAddonsFilterSchema);
```
```suggestion
export const EksAddonsFilter = Json.pipe(EksAddonsFilterSchema);
```

`Json` needs to be imported from our schema utils.

viceice commented on lib/modules/datasource/aws-eks-addon/index.ts:
```diff
@@ -0,0 +1,70 @@
+import { cache } from '../../../util/cache/package/decorator';
+import { Datasource } from '../datasource';
+import type { GetReleasesConfig, ReleaseResult } from '../types';
+import { EksAddonsFilterJson } from './schema';
```
```suggestion
import { EksAddonsFilter } from './schema';
```

viceice commented on lib/modules/datasource/aws-eks-addon/index.ts:
```diff
@@ -0,0 +1,70 @@
+  async getReleases({
+    packageName: serializedFilter,
+  }: GetReleasesConfig): Promise<ReleaseResult | null> {
+    const filter = EksAddonsFilterJson.safeParse(serializedFilter);
```
```suggestion
    const filter = EksAddonsFilter.safeParse(serializedFilter);
```
